include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -680,44 +682,44 @@ struct __copy2_wrapper;
 
 template <typename _BackendTag, typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3,
           typename _Compare>
-oneapi::dpl::__internal::__difference_t<_Range3>
+std::pair<oneapi::dpl::__internal::__difference_t<_Range1>, oneapi::dpl::__internal::__difference_t<_Range2>>
 __pattern_merge(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _Range1&& __rng1, _Range2&& __rng2,
                 _Range3&& __rng3, _Compare __comp)
 {
     auto __n1 = __rng1.size();
     auto __n2 = __rng2.size();
-    auto __n = __n1 + __n2;
-    if (__n == 0)
-        return 0;
+    if (__rng3.size() == 0)
+        return {0, 0};
 
     //To consider the direct copying pattern call in case just one of sequences is empty.
     if (__n1 == 0)
```

Comment:
We can make additional optimization here for the case when last(rng1) < first(rng2)

Reply:
> the case when last(rng1) < first(rng2)

As far as I remember the standard tells that such case is UB. 
In general  we don't introduce such checks. But in cases where we already have zero/empty checks we probably can extend the checks, like `if (__n == 0)`  -> `if (__n < 0)` or `if(first == last)` -> `if(first >= last)`. We can discuss it with the team.

Reply:
"As far as I remember the standard tells that such case is UB." - why?
For example, rng1 is 1, 2, 3 and rng2 is 4, 5, 6
This data is OK but merge as process don't needed in this case.

Reply:
Discussed offline, closed.